tests/realtikvtest/txntest/stale_read_test.go

```diff
@@ -1506,3 +1506,210 @@ func TestStaleReadNoBackoff(t *testing.T) {
+	firstTime := time.Now().Add(-250 * time.Millisecond)
+	// Retrieve current TSO from store's Oracle instead of @@tidb_current_ts.
+	externalTS, err := store.GetOracle().GetTimestamp(context.Background(), &oracle.Option{})
+	if err != nil {
+		t.Fatalf("failed to get TSO: %v", err)
+	}
+
+	time.Sleep(500 * time.Millisecond)
+	// Insert row #2
+	tk.MustExec("insert into t values (2, 20)")
+	time.Sleep(500 * time.Millisecond)
+	// Record approximate insert time for row #2 (offset: 250ms)
+	secondTime := time.Now().Add(-250 * time.Millisecond)
+
+	staleReadMethods := []struct {
+		name   string
+		setup  func()
+		query  string
+		clean  func()
+		expect []string
+	}{
+		{
+			name: "tidb_read_staleness",
+			setup: func() {
+				tk.MustExec("set @@tidb_read_staleness='-1'")
+			},
+			query: "select * from t",
+			clean: func() {
+				tk.MustExec("set @@tidb_read_staleness=''")
+			},
+			expect: []string{"1 10"},
+		},
+		{
+			name:   "AS OF TIMESTAMP sees only first row",
+			setup:  func() {},
+			query:  fmt.Sprintf("SELECT * FROM t AS OF TIMESTAMP '%s'", firstTime.Format("2006-1-2 15:04:05.000")),
+			clean:  func() {},
+			expect: []string{"1 10"},
+		},
+		{
+			name:   "AS OF TIMESTAMP sees both rows (using secondTime)",
+			setup:  func() {},
+			query:  fmt.Sprintf("SELECT * FROM t AS OF TIMESTAMP '%s'", secondTime.Format("2006-1-2 15:04:05.000")),
+			clean:  func() {},
+			expect: []string{"1 10", "2 20"},
+		},
+		/*
+			We cannot test TIDB_BOUNDED_STALENESS here because it relies
+			on the update of safe ts, which cannot be controlled by the test
+			{
+				name:   "AS OF TIMESTAMP with TIDB_BOUNDED_STALENESS",
+				setup:  func() {},
+				query:  "select * from t as of timestamp tidb_bounded_staleness(TIMESTAMPADD(SECOND, -5, NOW()), TIMESTAMPADD(SECOND, -1, NOW()))",
+				clean:  func() {},
+				expect: []string{"1 10", "2 20"},
+			},
+		*/
+		{
+			name: "SET TRANSACTION",
+			setup: func() {
+				tk.MustExec(fmt.Sprintf("SET TRANSACTION READ ONLY AS OF TIMESTAMP '%s'", firstTime.Format("2006-1-2 15:04:05.000")))
+			},
+			query:  "select * from t",
+			clean:  func() {},
+			expect: []string{"1 10"},
+		},
+		{
+			name: "external ts read",
+			setup: func() {
+				tk.MustExec("set @@tidb_enable_external_ts_read=1")
+				// ignore the failure when setting external_ts, as it reports error if not set to a larger value
+				_, _ = tk.Exec(fmt.Sprintf("set @@global.tidb_external_ts=%d", externalTS))
+			},
+			query: "select * from t",
+			clean: func() {
+				tk.MustExec("set @@tidb_enable_external_ts_read=0")
+				// cannot set it
+				// tk.MustExec("set @@global.tidb_external_ts=0")
+			},
+			expect: []string{"1 10"},
+		},
+	}
+
+	labelSettings := []struct {
+		name   string
+		labels map[string]string
+	}{
+		{
+			name:   "no labels",
+			labels: map[string]string{},
+		},
+		{
+			name: "with DC label",
+			labels: map[string]string{
+				placement.DCLabelKey: "bj",
+			},
+		},
```

Comment:
The `zone`, `dc`, `rack` could also be covered though some may not take effect on closest.